### PR TITLE
network: optionally bring up interface before joining bridge

### LINKED ISSUE #34247: Using systemd-networkd with TI switchdev switch
I'm using a chip from TI which contains a hardware switch. The switch driver uses the switchdev interface. So the switch can be configured via the bridge device.

The TI documentation has the requirement for the switch ports to be UP before joining the bridge.

"Port’s netdev devices have to be in UP state before joining the bridge"

[TI CPSW](https://software-dl.ti.com/jacinto7/esd/processor-sdk-linux-j721s2/latest/exports/docs/linux/Foundational_Components/Kernel/Kernel_Drivers/Network/CPSWng-Native-Ethernet.html#switch-mode)

Can this somehow be configured with systemd-networkd?

Currently, my switch ports go UP after joining the bridge. This leads to a wrong configuration. This can be obsevered in the latest systemd release and older version as well.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #34247.

### CODE REVIEW

File: rules.d/81-net-bridge.rules [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="net_bridge_end"
+SUBSYSTEM!="net", GOTO="net_bridge_end"
+
+# Some devices require the port is up before joining the bridge.
```
Comment (dnicolodi) [inline rules.d/81-net-bridge.rules]: `s/the port is up/the port to be up/`

File: rules.d/81-net-bridge.rules [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+# In such cases, ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE to "1".
```
Comment (dnicolodi) [inline rules.d/81-net-bridge.rules]: `s/In such cases, ID_/In such cases, set ID_/`

**Comment by yuwata** [comment]
@dnicolodi Thank you for the comments. Addressed the suggestions.

**Comment by poettering** [comment]
Hmm, do we have any feedback from relevant kernel networking folks that this really is the right way to go? i.e. is this a fundamental, expected, good behaviour here? Or is this just some limitation of a specific driver that behaves differently than it should and where there is a perspective that the driver is just fixed?

anyone has contacts to the maintainers of this driver and its subsystem?

**Comment by Siddharth-Vadapalli-at-TI** [comment]
Hello @poettering,

Please consider the following use-case:
An Ethernet Switch has two MAC ports with the corresponding Linux network interfaces being eth0 and eth1. Additionally, the Ethernet Switch forwards traffic _selectively_ (as explained below) across the MAC ports, while also sending traffic to/from the Host (CPU) port for software to consume/transmit. The VLAN configuration for the MAC Ports (or equivalently their network interfaces) is as follows:
eth0 => Belongs to VLAN A, VLAN B
eth1 => Belongs to VLAN B, VLAN C
The VLAN configuration for the Host (CPU) port (represented by the bridge interface br0) is as follows:
br0 => Belongs to VLAN A, VLAN B, VLAN C

Additionally, please note the following:
1. Rules/Configuration on the bridge can be applied to the underlying interfaces as well
2. **There is no rules/configuration replay mechanism for the bridge interface**
3. Commands/Rules executed on interfaces when they are down will not update the ALE (Address Lookup Engine) and the Switch configuration w.r.t. the MAC Ports the interfaces represent, due to the understanding that an interface being down means that there is no link.

Therefore, an interface being UP affects whether or not the configuration performed on it takes effect. At the same time, this also implies that an interface being set under a bridge affects whether or not the confiugration performed on the bridge is offloaded to the interface.

For the use-case mentioned above, one possible working sequence is:
1. Bring up eth0 and eth1
2. Setup VLANs A and B on eth0 and VLANs B and C on eth1
3. Bring up br0 and add eth0 and eth1 as the underlying interfaces
4. Setup VLANs A, B and C only on br0 (no offloading)
Another possible working sequence is:
1. Bring up eth0, br0, with eth0 and eth1 added as the underlying interfaces for br0
2. Setup VLAN A on eth0 and br0, VLAN C on eth1 and br1, VLAN B on br0 (offloaded to eth0 and eth1)

A sequence which will not work is:
1. eth0 and eth1 are added as the underlying interfaces for br0 (but not brought up)
2. Setup VLAN A on eth0 and br0, VLAN C on eth1 and br1, VLAN B on br0 (offloaded to eth0 and eth1)
3. Bring up eth0 and eth1 => Not only will it be the case that eth0 doesn't belong to VLAN A and eth1 doesn't belong to VLAN C, but because br0 will not replay VLAN B command, eth0 and eth1 won't belong to VLAN B as well.

In brief, since the interface state affects whether or not a configuration will take effect, it will be helpful to control the sequence of interface being brought up and the configuration being applied to an interface.

Regards,
Siddharth.

**Comment by matthoe** [comment]
Is the explanation above sufficient or do you need more to merge this PR?

**Comment by Siddharth-Vadapalli-at-TI** [comment]
Hello @poettering,

Sorry for the confusion. I am summarizing the driver limitation and the requirement below:

- The CPSW driver supports toggling between EMAC and Switch mode at runtime (a feature not supported by most if not all Ethernet drivers in Linux)
- This feature comes at the cost of the limitation that the interfaces should be up before joining the bridge in order to avoid overwriting the bridge configuration
- Based on the discussions on the Linux Mailing lists dating back to 2019, the limitation was accepted since it simplifies the driver code
- The requirement therefore is that of having the ability to perform actions on interfaces (toggling state of interfaces in this case) before they are bound to a bridge or any other networking interface

The requirement stated above helps overcome the limitation of the CPSW driver, but it solves a part of the generic requirement where users can translate a sequence of userspace networking commands into systemd rules.

For example, the following sequence:
ip link set eth0 up
ip link set eth1 up
ip link add br0 type bridge
ip link set eth0 master br0
ip link set eth1 master br0
cannot be guaranteed with the current systemd rules. So features that are functional when userspace networking commands are used, may break when users attempt to translated them into a systemd configuration.

Regards,
Siddharth.

**Comment by yuwata** [comment]
@matthoe @Siddharth-Vadapalli-at-TI 
From https://software-dl.ti.com/jacinto7/esd/processor-sdk-linux-j721s2/latest/exports/docs/linux/Foundational_Components/Kernel/Kernel_Drivers/Network/CPSWng-Native-Ethernet.html#switch-mode, such 'workaround' is required only when the switch mode is enabled.

How can we get if the switch mode is enabled ?
I guess something like the following:
`devlink dev param show platform/c000000.ethernet name switch_mode`
Is this correct?

How can we get the name of the devlink device `platform/c000000.ethernet` from the network interface name e.g. `eth0`?
Could you show the result of `udevadm info /sys/class/net/eth0` ?

**Comment by yuwata** [comment]
The devlink command line can be used for other drivers?
E.g. https://edc.intel.com/content/www/xl/es/design/products/ethernet/appnote-e810-eswitch-switchdev-mode-config-guide/configuration-steps/
The command line seems to be slightly different, and no requirement about the bridge up/down state is mentioned.

**Comment by yuwata** [comment]
I think the current form of the change to networkd should be OK. Potentially, there may exist other drivers with the similar restriction.
But, actually, udev rules here should be more specific and the flag should be set only when the switch_mode is enabled.

So, please let me know the way to get the current switch_mode setting, especially the way to obtain the device name like 'platform/c000000.ethernet' from `eth0`.

**Comment by Siddharth-Vadapalli-at-TI** [comment]
Hello @yuwata,

Devlink is used by other drivers as well and devlink is not limited to Switch Mode. It may be used for other features depending on the driver.

To view the current value of the switch_mode parameter, as you have rightly pointed out, the following command can be run:
`devlink dev param show platform/c000000.ethernet name switch_mode`
The output is:
`platform/c000000.ethernet:
  name switch_mode type driver-specific
    values:
      cmode runtime value false
`

The device name `platform/c000000.ethernet` can be obtained via ethtool by running the following command:
`ethtool -i ethX`
with <ethX> replaced by the corresponding interface. The output is:
`
driver: am65-cpsw-nuss
version: 6.6.44-01388-gfc249e610bc1
firmware-version: 
expansion-rom-version: 
bus-info: c000000.ethernet
supports-statistics: yes
supports-test: no
supports-eeprom-access: no
supports-register-dump: yes
supports-priv-flags: yes
`
with the "bus-info" field containing the parameter of interest.

Regards,
Siddharth.

**Comment by Siddharth-Vadapalli-at-TI** [comment]
While I don't have the scripts being used by @matthoe, when I run the command:
`udevadm info /sys/class/net/eth1`
the output is:
`P: /devices/platform/bus@100000/c000000.ethernet/net/eth1`
`M: eth1`
`R: 1`
`U: net`
`I: 3`
`E: DEVPATH=/devices/platform/bus@100000/c000000.ethernet/net/eth1`
`E: SUBSYSTEM=net`
`E: INTERFACE=eth1`
`E: IFINDEX=3`
`E: USEC_INITIALIZED=9091273`
`E: ID_NET_DRIVER=am65-cpsw-nuss`
`E: ID_NET_NAMING_SCHEME=v255`
`E: ID_NET_NAME_MAC=enx70ff761dff79`
`E: ID_OUI_FROM_DATABASE=Texas Instruments`
`E: ID_PATH=platform-c000000.ethernet`
`E: ID_PATH_TAG=platform-c000000_ethernet`
`E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link`
`E: ID_NET_NAME=eth1`
`E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/eth1`
`E: TAGS=:systemd:`
`E: CURRENT_TAGS=:systemd:`

**Comment by yuwata** [comment]
@Siddharth-Vadapalli-at-TI 
Thanks. Could you also show
`udevadm info /sys/devices/platform/bus@100000/c000000.ethernet` ?

**Comment by Siddharth-Vadapalli-at-TI** [comment]
@yuwata, the output of:
`udevadm info /sys/devices/platform/bus@100000/c000000.ethernet`
on J721E-EVM is:
`P: /devices/platform/bus@100000/c000000.ethernet`
`M: c000000.ethernet`
`U: platform`
`V: am65-cpsw-nuss`
`E: DEVPATH=/devices/platform/bus@100000/c000000.ethernet`
`E: SUBSYSTEM=platform`
`E: DRIVER=am65-cpsw-nuss`
`E: OF_NAME=ethernet`
`E: OF_FULLNAME=/bus@100000/ethernet@c000000`
`E: OF_COMPATIBLE_0=ti,j721e-cpswxg-nuss`
`E: OF_COMPATIBLE_N=1`
`E: MODALIAS=of:NethernetT(null)Cti,j721e-cpswxg-nuss`
`E: USEC_INITIALIZED=9070987`
`E: ID_PATH=platform-c000000.ethernet`
`E: ID_PATH_TAG=platform-c000000_ethernet`

**Comment by yuwata** [comment]
@Siddharth-Vadapalli-at-TI @matthoe Updated the udev rule to make the flag only set when the switch mode is enabled. Could you test the PR?

@poettering WDYT about the current way?

File: rules.d/81-net-bridge.rules [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="net_bridge_end"
+
+# Some devices require the port to be up before joining the bridge.
+# In such cases, set ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE to "1".
+
+# Texas Instruments Ethernet device with switchdev mode:
+# https://software-dl.ti.com/jacinto7/esd/processor-sdk-linux-j721s2/latest/exports/docs/linux/Foundational_Components/Kernel/Kernel_Drivers/Network/CPSWng-Native-Ethernet.html#switch-mode
```
Comment (Siddharth-Vadapalli-at-TI) [inline rules.d/81-net-bridge.rules]: @yuwata, here and elsewhere, I suggest replacing the link with:
https://docs.kernel.org/networking/device_drivers/ethernet/ti/am65_nuss_cpsw_switchdev.html#enabling-switch
which is the Linux Documentation for the driver mentioning the same requirement/limitation of the driver.

**Review by Siddharth-Vadapalli-at-TI — APPROVED** [review]
Thank you @yuwata for updating the links. @matthoe, please test the PR.

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
@matthoe Fixed the udev rules. Could you test again? If it still does not work. Please provide the information above.

**Comment by matthoe** [comment]
Unfortunately it's still not working.

```
$ udevadm info /sys/class/net/xf4
P: /devices/platform/bus@100000/c000000.ethernet/net/xf4
M: xf4
R: 4
U: net
I: 8
E: DEVPATH=/devices/platform/bus@100000/c000000.ethernet/net/xf4
E: SUBSYSTEM=net
E: INTERFACE=xf4
E: IFINDEX=8
E: USEC_INITIALIZED=4845062
E: ID_NET_NAMING_SCHEME=v252
E: ID_PATH=platform-c000000.ethernet
E: ID_PATH_TAG=platform-c000000_ethernet
E: ID_NET_DRIVER=am65-cpsw-nuss
E: ID_NET_LINK_FILE=/etc/systemd/network/80-switch.link
E: ID_NET_NAME=xf4
E: ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE=1
E: ID_BOARD_NAME=krc5-basic2
E: ID_BOARD_SOC=gp
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/xf4
E: TAGS=:systemd:
E: CURRENT_TAGS=:systemd:
```

```
$ udevadm info /devices/platform/bus@100000/c000000.ethernet
Unknown device "/devices/platform/bus@100000/c000000.ethernet": No such device
```

```
$ udevadm info /sys/devices/platform/bus@100000/c000000.ethernet
P: /devices/platform/bus@100000/c000000.ethernet
M: c000000.ethernet
U: platform
V: am65-cpsw-nuss
E: DEVPATH=/devices/platform/bus@100000/c000000.ethernet
E: SUBSYSTEM=platform
E: DRIVER=am65-cpsw-nuss
E: OF_NAME=ethernet
E: OF_FULLNAME=/bus@100000/ethernet@c000000
E: OF_COMPATIBLE_0=ti,j784s4-cpswxg-nuss
E: OF_COMPATIBLE_N=1
E: MODALIAS=of:NethernetT(null)Cti,j784s4-cpswxg-nuss
E: USEC_INITIALIZED=3900707
E: ID_PATH=platform-c000000.ethernet
E: ID_PATH_TAG=platform-c000000_ethernet
```

```
$ devlink dev param show platform/c000000.ethernet name switch_mode
platform/c000000.ethernet:
  name switch_mode type driver-specific
    values:
      cmode runtime value true
```

**Comment by yuwata** [comment]
Hm? Why `ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE` is set without `ID_NET_DEVLINK_BUS_NAME` ?
Did you applied the patch cleanly and reboot the system?

**Comment by yuwata** [comment]
How about the revised one?

If it still does not work, please create the following udev rule:
```
# /etc/udev/rules.d/00-net-debug.rules
SUBSYSTEM=="net", OPTIONS="log_level=debug"
```
then reboot the system, and provides the debugging logs of systemd-udevd
```
journalctl -b --no-hostname -u systemd-udevd.service
```
Also, as usual, please provide the result of `udevadm info /sys/class/net/xf4`.

**Comment by matthoe** [comment]
> Hm? Why `ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE` is set without `ID_NET_DEVLINK_BUS_NAME` ? Did you applied the patch cleanly and reboot the system?

Sorry for the confusion. I mixed up the outputs from the different udev rules. There was no `ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE` with that version of the udev rule.


The current version of the udev rule is working for me. 

```
# do not edit this file, it will be overwritten on update

ACTION=="remove", GOTO="net_bridge_end"
SUBSYSTEM!="net", GOTO="net_bridge_end"

# Some devices require the port to be up before joining the bridge.
# In such cases, set ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE to "1".

# Texas Instruments Ethernet device with switchdev mode:
# https://docs.kernel.org/networking/device_drivers/ethernet/ti/am65_nuss_cpsw_switchdev.html#enabling-switch
ENV{ID_NET_DRIVER}=="am65-cpsw-nuss", SUBSYSTEMS=="platform", DRIVERS=="am65-cpsw-nuss", \
  PROGRAM="/usr/sbin/devlink dev param show platform/%b name switch_mode", \
  RESULT=="*cmode runtime value true*", \
  ENV{ID_NET_BRING_UP_BEFORE_JOINING_BRIDGE}="1"

LABEL="net_bridge_end"
```

Previously, the program failed with
```
(udev-worker)[315]: eth0: Process 'devlink dev param show platform/c000000.ethernet name switch_mode' failed with exit code 1.
(udev-worker)[315]: eth0: /lib/udev/rules.d/81-net-bridge.rules:11 Command "devlink dev param show platform/c000000.ethernet name switch_mode" returned 1 (error)
```

`/usr/sbin/devlink` fixed the issue.

**Comment by yuwata** [comment]
@matthoe Thank you for testing so many times!

**Pull request merged**